src/Compilers/CSharp/Portable/Binder/Binder_NameConflicts.cs

```diff
@@ -65,16 +65,37 @@ internal void ValidateParameterNameConflicts(
                         continue;
                     }
 
-                    if (tpNames != null && tpNames.Contains(name))
+                    if (tpNames != null && tpNames.TryGetValue(name, out TypeParameterSymbol? tp))
                     {
-                        // CS0412: 'X': a parameter or local variable cannot have the same name as a method type parameter
-                        diagnostics.Add(ErrorCode.ERR_LocalSameNameAsTypeParam, GetLocation(p), name);
+                        if (tp.ContainingSymbol is NamedTypeSymbol { IsExtension: true })
+                        {
+                            if (p.ContainingSymbol != (object)tp.ContainingSymbol)
```

Comment:
nit: consider leaving a comment (the error for naming the extension parameter after an extension type parameter is reported elsewhere)